plugins/GranularPitchShifter/GranularPitchShifterEffect.cpp

```diff
@@ -0,0 +1,269 @@
+	{
+		const double pitch = pitchBuf ? pitchBuf->value(f) : m_granularpitchshifterControls.m_pitchModel.value();
+		const double pitchSpread = (pitchSpreadBuf ? pitchSpreadBuf->value(f) : m_granularpitchshifterControls.m_pitchSpreadModel.value()) * 0.5f;
+		
+		// interpolate pitch depending on glide
+		for (int i = 0; i < 2; ++i)
+		{
+			double targetVal = pitch + pitchSpread * (i ? 1. : -1.);
+			
+			if (targetVal == m_truePitch[i]) { continue; }
+			m_updatePitches = true;
+			
+			m_truePitch[i] = m_glideCoef * m_truePitch[i] + (1. - m_glideCoef) * targetVal;
+			// we crudely lock the pitch to the target value once it gets close enough, so we can save on CPU
+			if (std::abs(targetVal - m_truePitch[i]) < GlideSnagRadius) { m_truePitch[i] = targetVal; }
+		}
+		
+		// this stuff is computationally expensive, so we should only do it when necessary
+		if (m_updatePitches)
+		{
+			m_updatePitches = false;
+			
+			std::array<double, 2> speed = {std::exp2(m_truePitch[0] * (1. / 12.)), std::exp2(m_truePitch[1] * (1. / 12.))};
+			std::array<double, 2> ratio = {speed[0] / m_speed[0], speed[1] / m_speed[1]};
+			
+			for (int i = 0; i < m_grainCount; ++i)
+			{
+				for (int j = 0; j < 2; ++j)
+				{
+					m_grains[i].grainSpeed[j] *= ratio[j];
+					
+					// we unfortunately need to do extra stuff to ensure these don't shoot past the write index...
+					if (m_grains[i].grainSpeed[j] > 1)
+					{
+						double distance = m_writePoint - m_grains[i].readPoint[j] - SafetyLatency;
+						if (distance <= 0) { distance += m_ringBufLength; }
+						double grainSpeedRequired = ((m_grains[i].grainSpeed[j] - 1.) / distance) * (1. - m_grains[i].phase);
+						m_grains[i].phaseSpeed[j] = std::max(m_grains[i].phaseSpeed[j], grainSpeedRequired);
+					}
+				}
+			}
+			m_speed[0] = speed[0];
+			m_speed[1] = speed[1];
+			
+			// prevent aliasing by lowpassing frequencies that the pitch shifting would push above nyquist
+			m_prefilter[0].setCoefs(m_sampleRate, std::min(m_nyquist / static_cast<float>(speed[0]), m_nyquist) * PrefilterBandwidth);
+			m_prefilter[1].setCoefs(m_sampleRate, std::min(m_nyquist / static_cast<float>(speed[1]), m_nyquist) * PrefilterBandwidth);
+		}
+		
+		std::array<float, 2> s = {0, 0};
+		std::array<float, 2> filtered = {buf[f][0], buf[f][1]};
+		
+		// spawn a new grain if it's time
+		if (++m_timeSinceLastGrain >= m_nextWaitRandomization * waitMult)
+		{
+			m_timeSinceLastGrain = 0;
+			double randThing = (fast_rand()/static_cast<double>(FAST_RAND_MAX) * 2. - 1.);
+			m_nextWaitRandomization = std::exp2(randThing * twitch);
+			double grainSpeed = 1. / std::exp2(randThing * jitter);
+
+			std::array<float, 2> sprayResult = {0, 0};
+			if (spray > 0)
+			{
+				sprayResult[0] = (fast_rand() / static_cast<float>(FAST_RAND_MAX)) * spray * m_sampleRate;
+				sprayResult[1] = linearInterpolate(sprayResult[0],
+									(fast_rand() / static_cast<float>(FAST_RAND_MAX)) * spray * m_sampleRate,
+									spraySpread);
```

Comment:
Less indent heavy here
```suggestion
				sprayResult[1] = linearInterpolate(
					sprayResult[0],
					(fast_rand() / static_cast<float>(FAST_RAND_MAX)) * spray * m_sampleRate,
					spraySpread
				);
```